fix(file): UseBlob upload path double-closes src, causing Windows WebFTP downloads to always be reported as failed

## Summary

Since the v2.1.6 release, downloading a file via WebFTP on a Windows host (from the browser's perspective, which maps to alpamon's `upload` command) results in alpamon reporting `exitCode=1, "Failed to upload file"` to alpacon-server even though the underlying PUT reaches alpacon-server successfully. Users see the operation marked as failed in the console UI and repeatedly retry the same file.

Operator-reported log:

```
2026-05-21T09:48:52Z [ERROR] Failed to upload file
  error="close C:\\Users\\Administrator\\Desktop\\AlpacaX_2026_Strategic_Roadmap_Confidential_KO.pdf: file already closed"
2026-05-21T09:48:52Z [INFO] Command executed successfully (command)upload (duration)380.7409 (exitCode)1
```

The transfer itself completed over 380.74 seconds and the PUT returned a 2xx response. The core of this regression is that alpamon's `src.Close()` call right after the PUT returns `os.ErrClosed`, which is then mistakenly converted into an upload failure. The message misleads operators toward disk handle, antivirus, or file-locking tooling, but the actual root cause is that alpamon closes the same handle twice, ignoring `http.Client.Do`'s automatic-close contract.

## Observed behavior

### Consistency of symptoms

- On Windows, **every** UseBlob upload path is reported as failed identically (because they all go through the `readFileAs` branch at [`pkg/executor/handlers/file/file_io_windows.go:13`](pkg/executor/handlers/file/file_io_windows.go#L13), which returns `*os.File` as the source).
- Independent of file size, extension, and content type (the `UseBlob` branch uses size only as the Content-Length hint for the PUT; close behavior is unaffected).
- The same file is correctly stored in alpacon-server's object storage (the PUT itself succeeds).
- The symptom looks more dramatic on long durations (large files) but actually occurs for all sizes.

### Differences on Linux/macOS

On Linux/macOS, `readFileAs` returns a `cmdReadCloser` wrapping a demoted `cat` subprocess ([`pkg/executor/handlers/file/file_io_unix.go:33`](pkg/executor/handlers/file/file_io_unix.go#L33)). That type's `Close()` is written to cache and re-return the cat exit code, so it behaves effectively idempotently — which is why this regression has been reported on Windows only. However, the same double-close belongs to the class of bugs that can trigger fd-reuse races (Go issue #39907 family), so it should be treated as a latent defect on Unix as well.

### Exit code contradiction

The most striking diagnostic signal is that the `Command executed successfully` INFO log and `exitCode=1` appear on the same line. `handleUpload` ([`pkg/executor/handlers/file/file.go:173-177`](pkg/executor/handlers/file/file.go#L173-L177)) logs `Failed to upload file` and returns `(1, err.Error())`, but the upstream dispatcher still prints an INFO line meaning "the command itself ran to completion". Operators read this as "so did it succeed or fail?".

## Root cause

The cause is that alpamon `fileUpload`'s `UseBlob` branch **duplicates** `http.Client.Do`'s automatic-close contract.

### alpamon side (since v2.1.6, [`pkg/executor/handlers/file/file.go:397-407`](pkg/executor/handlers/file/file.go#L397-L407))

```go
func (h *FileHandler) fileUpload(args *common.CommandArgs, src io.ReadCloser, size int64, fileName string, recursive bool) (int, error) {
    if args.UseBlob {
        _, code, err := utils.Put(args.Content, src, size, 0)
        // Surface src.Close() errors when the PUT itself succeeded so a
        // non-zero `cat` exit (cmdReadCloser EACCES/ENOENT) is not silently
        // reported as a successful upload.
        if cerr := src.Close(); err == nil && cerr != nil {
            return code, cerr
        }
        return code, err
    }
    ...
}
```

The intent is correct — surface the Unix `cmdReadCloser`'s non-zero cat exit (EACCES/ENOENT) so it isn't buried as a silent success. The result is a single line calling `src.Close()` to capture the cat exit code.

### `net/http` side (Go standard library contract)

[`utils.Put`](pkg/utils/http_client.go#L45) passes src through to the body via `http.NewRequest(http.MethodPut, url, body)`. From the Go `net/http` documentation for `func NewRequest`:

> If the provided body is also an io.Closer, the returned Request.Body is set to body and will be closed (possibly asynchronously) by the Client methods Do, Post, and PostForm, and Transport.RoundTrip.

So when `src` is an `io.Closer` (both `*os.File` and `cmdReadCloser` are), `req.Body = src` is set directly and `req.Body.Close()` is called automatically when `client.Do(req)` finishes. **This is where the first close happens.**

### Double-close flow

| Step | Call site | Behavior |
|---|---|---|
| 1 | `readFileAs` ([`file_io_windows.go:13`](pkg/executor/handlers/file/file_io_windows.go#L13)) | Returns `*os.File` |
| 2 | `utils.Put` → `http.NewRequest` | `req.Body = src` (recognized as Closer, used as-is) |
| 3 | `utils.Put` → `client.Do(req)` | After sending body, calls `req.Body.Close()` automatically → **first close (succeeds)** |
| 4 | `fileUpload` line 403 | Calls `src.Close()` → `*os.File` is not idempotent, returns `&PathError{Op:"close", Path:..., Err:os.ErrClosed}` |
| 5 | `fileUpload` line 403-405 | `err == nil && cerr != nil` matches → returns `(code, cerr)` |
| 6 | `handleUpload` ([`file.go:173-177`](pkg/executor/handlers/file/file.go#L173-L177)) | Logs "Failed to upload file" + exit 1 |

A second call to `os.File.Close()` is explicitly defined by the Go standard library to throw `ErrClosed`, and the message format is exactly `close <path>: file already closed` — character-for-character identical to what the operator observed.

### How it was introduced

The intent stated in the commit message — "stop the demoted cat's EACCES/ENOENT from being buried as silent success" — is legitimate. But it overlooked the fact that `http.Client.Do` already calls `Body.Close()` (and ignores its returned error). As a result, the intent (surfacing cat errors) is only half achieved while Windows is hit by a guaranteed false positive on every upload.

## Reproduction

On a clean Windows host, install alpamon v2.1.6 or v2.1.7, register, and try downloading any file (preferably ≥ 1 MB) through WebFTP from the console:

1. Register and start:
   ```powershell
   & "<path>\alpamon.exe" register --url <ALPACON_URL> --token <TOKEN>
   sc.exe start alpamon
   ```
2. Alpacon web console → registered server → WebFTP tab → download any file (e.g. `C:\Users\Administrator\Desktop\anything.pdf`).
3. Check alpamon-side logs (remote log channel or stderr capture):
   ```
   [ERROR] Failed to upload file error="close <path>: file already closed"
   [INFO] Command executed successfully (command)upload ... (exitCode)1
   ```
4. Check the task state on alpacon-server → shown as failed, but the actual object is stored.

Smaller files that take the multipart branch (`UseBlob=false`) are not affected. This regression applies to every case where the console issues a command with `UseBlob=true` — typically every WebFTP download.

## Workaround (operator-side, until fix ships)

There is effectively no operator-side workaround — alpamon reports failure, but the file itself reaches alpacon-server, so the only client-side option is to ignore the failure indicator in the console UI and fetch the file by another path from alpacon-server's object storage. This regression can only be fixed by an alpamon-side patch.

Therefore:
- An immediate patch release (v2.1.8) must be cut.
- In the meantime, an operational notice along the lines of "Windows WebFTP downloads marked as failed are actually preserved server-side; ignore the task-state UI and obtain the file via the direct download link until the patch ships" is needed — though this depends on alpacon-server's object retention policy and must be agreed with the server team first.

## Proposed change

Three candidate fixes, in recommended order:

### Option A (recommended): strip the Closer face with `io.NopCloser`

```diff
 if args.UseBlob {
-    _, code, err := utils.Put(args.Content, src, size, 0)
-    // Surface src.Close() errors when the PUT itself succeeded so a
-    // non-zero `cat` exit (cmdReadCloser EACCES/ENOENT) is not silently
-    // reported as a successful upload.
+    // Strip Closer so http.Client.Do can't close src before us — we own
+    // the close so the demoted-cat reader's non-zero exit (EACCES/ENOENT)
+    // surfaces via Close(), and *os.File on Windows is not double-closed
+    // (returns os.ErrClosed on the second Close, which would otherwise be
+    // reported as a failed upload even though the PUT succeeded).
+    _, code, err := utils.Put(args.Content, io.NopCloser(src), size, 0)
     if cerr := src.Close(); err == nil && cerr != nil {
         return code, cerr
     }
     return code, err
 }
```

- `http.NewRequest` still uses the NopCloser wrapper as `req.Body`, but its Close is a no-op, so `client.Do`'s automatic close effectively disappears.
- The explicit `src.Close()` becomes the single real close, and the demoted-cat exit code is surfaced exactly as originally intended.
- Change footprint: 1 line (the wrapping) + 4 lines of comment. Semantics are clear and identical across Windows, Linux, and macOS.
- No external signature change; no impact on the caller (`handleUpload`).

### Option B: ignore `os.ErrClosed` only

```diff
-if cerr := src.Close(); err == nil && cerr != nil {
+if cerr := src.Close(); err == nil && cerr != nil && !errors.Is(cerr, os.ErrClosed) {
     return code, cerr
 }
```

- Suppresses the Windows symptom.
- However, **the double-close itself remains**. In the `*os.File` fd-reuse race class (Go issue #39907 family), if the closed fd is reallocated to another handle in between, our second Close could sever that handle — extremely rare, but leads to undiagnosable data corruption.
- Poor ratio between "symptom hidden" and "root cause unresolved".

### Option C: remove the explicit close

```diff
-_, code, err := utils.Put(args.Content, src, size, 0)
-if cerr := src.Close(); err == nil && cerr != nil {
-    return code, cerr
-}
-return code, err
+_, code, err := utils.Put(args.Content, src, size, 0)
+return code, err
```

- Simplest, but entirely abandons the intent behind `b9ba9712` (surfacing cmdReadCloser EACCES/ENOENT).
- `http.Client.Do` ignores the return value of `req.Body.Close()`, so the demoted-cat exit code is silently buried.
- The combination of an empty file (Content-Length 0) and a cat EACCES would be reported as a successful 0-byte upload.

### Conclusion

Option A strikes the best balance across change size, consistency, and comprehension cost. A 1-line wrapping preserves the intent (the cat-surface goal of `b9ba9712`) while removing both the Windows regression and the latent race.

## Acceptance criteria

- [ ] On Windows with alpamon v2.1.8+, performing a WebFTP download results in alpamon reporting `exitCode=0, "Successfully uploaded N file(s)."` for the same file.
- [ ] The message `close ... file already closed` no longer appears in alpamon logs.
- [ ] The WebFTP task state in the alpacon-server console is shown as success.
- [ ] On Linux/macOS, when demoted cat exits with EACCES/ENOENT (and `cmdReadCloser.Close()` returns a non-zero error), that error is still propagated to alpacon-server. In other words, the silent-success regression `b9ba9712` was originally fixing must not be reintroduced.
- [ ] On paths where the PUT itself fails (HTTP 5xx, network disconnect, etc.), the PUT error takes precedence over the `src.Close()` error in the returned value (verifying the `err == nil &&` guard).
- [ ] New unit tests lock in the three scenarios above (`TestFileUpload_UseBlob_OsFile_NoDoubleClose`, `TestFileUpload_UseBlob_CloseErrorPropagates`, `TestFileUpload_UseBlob_PutErrorTakesPrecedence`).
- [ ] No regression in the existing multipart path (`UseBlob=false`) — the full `pkg/executor/handlers/file` test suite passes.

## Out of scope

- Detailed audit of `cmdReadCloser.Close()` idempotency and fd-reuse safety on Linux/macOS. This patch is safe regardless of those assumptions, but reviewing the cat Wait()/Close() sequence is worth doing on a separate track.
- Comprehensive audit for the same anti-pattern (passing an `io.ReadCloser` directly to `http.NewRequest` and then having the caller Close again) at other call sites (`pkg/scheduler/transporter/`, `pkg/runner/ftp.go`, etc.). If found, split into a separate PR.
- How alpacon-server should reconcile cases where alpamon reported "failure" but the object was actually stored — tracked by the server team.
- Adding a WebFTP regression case (real PUT round-trip) to the GitHub Actions `windows-smoke-test`. It currently only verifies `--version`, which is why this regression slipped through. Split as an infrastructure task.
- Cleaning up the "Command executed successfully ... exitCode=1" contradiction in the log itself — once this fix lands, the exit code normalizes to 0 and the contradiction disappears naturally. Hardening the dispatcher to branch its INFO label on exit code is a separate cleanup.

## Reference signals for triage

To quickly match the same symptom when reported in the future:

- The message `[ERROR] Failed to upload file error="close ... file already closed"` in alpamon logs.
- Adjacent `[INFO] Command executed successfully (command)upload ... (exitCode)1` — the contradiction between INFO and exitCode 1.
- Even when the duration looks unusually long, traces show the PUT actually completed (the same key exists in alpacon-server's object storage).
- Windows operators reporting "downloads fail every time I retry" while alpacon-server's object storage accumulates duplicate entries.
- alpamon version v2.1.6 or v2.1.7 (tags containing commit `b9ba9712`).

## References

- Affected components:
  - [`pkg/executor/handlers/file/file.go`](pkg/executor/handlers/file/file.go#L397-L407) — the UseBlob branch of `fileUpload`.
  - [`pkg/executor/handlers/file/file_io_windows.go`](pkg/executor/handlers/file/file_io_windows.go#L13) — Windows `readFileAs` returns `*os.File`.
  - [`pkg/executor/handlers/file/file_io_unix.go`](pkg/executor/handlers/file/file_io_unix.go) — Unix `readFileAs` returns `cmdReadCloser`.
  - [`pkg/utils/http_client.go`](pkg/utils/http_client.go#L45) — the `utils.Put` call path.
- Introducing commit: `b9ba9712` "fix(file,utils): surface UseBlob src.Close errors and detect Put response truncation" (2026-05-07, PR #298).
- Affected releases: `v2.1.6-alpha.1`, `v2.1.6`, `v2.1.7`.
- Go `net/http` contract: `func NewRequest` doc — "If the provided body is also an io.Closer, ... will be closed ... by the Client methods Do, Post, and PostForm, and Transport.RoundTrip."
- Related latent risk: Go issue #39907 — double `os.File.Close()` calls and fd-reuse race.

--- a/pkg/executor/handlers/file/file.go
+++ b/pkg/executor/handlers/file/file.go
@@ -396,10 +396,12 @@ func (h *FileHandler) makeArchive(ctx context.Context, paths []string, bulk, rec
 //   - Pre-handoff failures: closes src directly so it never leaks.
 func (h *FileHandler) fileUpload(args *common.CommandArgs, src io.ReadCloser, size int64, fileName string, recursive bool) (int, error) {
 	if args.UseBlob {
-		_, code, err := utils.Put(args.Content, src, size, 0)
-		// Surface src.Close() errors when the PUT itself succeeded so a
-		// non-zero `cat` exit (cmdReadCloser EACCES/ENOENT) is not silently
-		// reported as a successful upload.
+		// Strip Closer so http.Client.Do can't close src before us—we own
+		// the close so the demoted-cat reader's non-zero exit (EACCES/ENOENT)
+		// surfaces via Close(), and *os.File on Windows is not double-closed
+		// (returns os.ErrClosed on the second Close, which would otherwise be
+		// reported as a failed upload even though the PUT succeeded).
+		_, code, err := utils.Put(args.Content, io.NopCloser(src), size, 0)
 		if cerr := src.Close(); err == nil && cerr != nil {
 			return code, cerr
 		}

--- a/pkg/executor/handlers/file/file_test.go
+++ b/pkg/executor/handlers/file/file_test.go
@@ -2,7 +2,14 @@ package file
 
 import (
 	"context"
+	"errors"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path/filepath"
 	"runtime"
+	"strings"
 	"testing"
 
 	"github.com/alpacax/alpamon/pkg/executor/handlers/common"
@@ -176,6 +183,99 @@ func TestFileExists(t *testing.T) {
 	}
 }
 
+// TestFileUpload_UseBlob_OsFile_NoDoubleClose locks in the v2.1.6 regression
+// where http.Client.Do auto-closes req.Body and fileUpload then calls
+// src.Close() a second time. On *os.File the second Close returns
+// os.ErrClosed, which fileUpload propagated as a failed upload. After the
+// io.NopCloser wrap, http.Client.Do can no longer close src and our
+// explicit Close is the single real close.
+func TestFileUpload_UseBlob_OsFile_NoDoubleClose(t *testing.T) {
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, _ = io.Copy(io.Discard, r.Body)
+		w.WriteHeader(http.StatusOK)
+	}))
+	defer srv.Close()
+
+	tmpPath := filepath.Join(t.TempDir(), "blob.bin")
+	if err := os.WriteFile(tmpPath, []byte("hello"), 0o600); err != nil {
+		t.Fatalf("write temp: %v", err)
+	}
+	f, err := os.Open(tmpPath)
+	if err != nil {
+		t.Fatalf("open temp: %v", err)
+	}
+
+	h := NewFileHandler(common.NewMockCommandExecutor(t), nil)
+	args := &common.CommandArgs{UseBlob: true, Content: srv.URL}
+
+	code, err := h.fileUpload(args, f, 5, "blob.bin", false)
+	if err != nil {
+		t.Fatalf("fileUpload returned err=%v, want nil (regression of v2.1.6 double-close)", err)
+	}
+	if code != http.StatusOK {
+		t.Errorf("fileUpload code=%d, want %d", code, http.StatusOK)
+	}
+}
+
+// TestFileUpload_UseBlob_CloseErrorPropagates verifies the original intent
+// of commit b9ba9712: when the underlying reader's Close() returns an
+// error (e.g. demoted cat EACCES/ENOENT via cmdReadCloser), fileUpload
+// must propagate it instead of reporting the upload as successful.
+// closeCnt==1 also guards against a future regression that brings the
+// double-close back.
+func TestFileUpload_UseBlob_CloseErrorPropagates(t *testing.T) {
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, _ = io.Copy(io.Discard, r.Body)
+		w.WriteHeader(http.StatusOK)
+	}))
+	defer srv.Close()
+
+	closeSentinel := errors.New("synthetic close failure")
+	er := &errReader{
+		r:        strings.NewReader("hello"),
+		failAt:   1 << 30,
+		closeErr: closeSentinel,
+	}
+
+	h := NewFileHandler(common.NewMockCommandExecutor(t), nil)
+	args := &common.CommandArgs{UseBlob: true, Content: srv.URL}
+
+	_, err := h.fileUpload(args, er, 5, "blob.bin", false)
+	if !errors.Is(err, closeSentinel) {
+		t.Fatalf("fileUpload err=%v, want chain containing %v", err, closeSentinel)
+	}
+	if er.closeCnt != 1 {
+		t.Errorf("errReader.Close was called %d time(s), want exactly 1 (double-close regression)", er.closeCnt)
+	}
+}
+
+// TestFileUpload_UseBlob_PutErrorTakesPrecedence verifies the `err == nil &&`
+// guard: when utils.Put itself fails (transport-level error), that error is
+// returned instead of the src.Close() error. Without this guard a Close()
+// failure would mask the real PUT failure.
+func TestFileUpload_UseBlob_PutErrorTakesPrecedence(t *testing.T) {
+	closeSentinel := errors.New("synthetic close failure")
+	er := &errReader{
+		r:        strings.NewReader("hello"),
+		failAt:   1 << 30,
+		closeErr: closeSentinel,
+	}
+
+	h := NewFileHandler(common.NewMockCommandExecutor(t), nil)
+	// 127.0.0.1:1 is the reserved tcpmux port; on all common platforms a
+	// connection here is refused immediately, producing a transport-level
+	// error from utils.Put without depending on DNS or external network.
+	args := &common.CommandArgs{UseBlob: true, Content: "http://127.0.0.1:1/blob"}
+
+	_, err := h.fileUpload(args, er, 5, "blob.bin", false)
+	if err == nil {
+		t.Fatal("fileUpload returned nil err, want PUT transport error")
+	}
+	if errors.Is(err, closeSentinel) {
+		t.Errorf("fileUpload returned close error %v; PUT transport error should take precedence", err)
+	}
+}
+
 func TestFileHandler_parsePaths(t *testing.T) {
 	// This test uses Unix-style absolute paths ("/home/user", "/tmp/...").
 	// On Windows those paths join with the supplied home into shapes
@@ -230,4 +330,3 @@ func TestFileHandler_parsePaths(t *testing.T) {
 		})
 	}
 }
-

--- a/pkg/executor/handlers/file/file_test.go
+++ b/pkg/executor/handlers/file/file_test.go
@@ -4,7 +4,6 @@ import (
 	"context"
 	"errors"
 	"io"
-	"net"
 	"net/http"
 	"net/http/httptest"
 	"os"
@@ -263,19 +262,14 @@ func TestFileUpload_UseBlob_PutErrorTakesPrecedence(t *testing.T) {
 	}
 
 	h := NewFileHandler(common.NewMockCommandExecutor(t), nil)
-	// Bind to 127.0.0.1:0 to claim a free port, then close the listener so
-	// any subsequent connection to that address is deterministically refused.
-	// This avoids depending on a specific port (e.g. tcpmux/1) being closed
-	// in CI/container environments.
-	ln, err := net.Listen("tcp", "127.0.0.1:0")
-	if err != nil {
-		t.Fatalf("listen: %v", err)
-	}
-	closedAddr := ln.Addr().String()
-	_ = ln.Close()
-	args := &common.CommandArgs{UseBlob: true, Content: "http://" + closedAddr + "/blob"}
+	// Port 0 is reserved and cannot be dialed; net.Dial fails immediately
+	// with "can't assign requested address" / "invalid argument", giving a
+	// deterministic transport-level error without depending on any port
+	// being closed or claiming an ephemeral port that could be re-bound
+	// in the window between listener close and the PUT attempt.
+	args := &common.CommandArgs{UseBlob: true, Content: "http://127.0.0.1:0/blob"}
 
-	_, err = h.fileUpload(args, er, 5, "blob.bin", false)
+	_, err := h.fileUpload(args, er, 5, "blob.bin", false)
 	if err == nil {
 		t.Fatal("fileUpload returned nil err, want PUT transport error")
 	}

--- a/pkg/executor/handlers/file/file_test.go
+++ b/pkg/executor/handlers/file/file_test.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"errors"
 	"io"
+	"net"
 	"net/http"
 	"net/http/httptest"
 	"os"
@@ -262,12 +263,19 @@ func TestFileUpload_UseBlob_PutErrorTakesPrecedence(t *testing.T) {
 	}
 
 	h := NewFileHandler(common.NewMockCommandExecutor(t), nil)
-	// 127.0.0.1:1 is the reserved tcpmux port; on all common platforms a
-	// connection here is refused immediately, producing a transport-level
-	// error from utils.Put without depending on DNS or external network.
-	args := &common.CommandArgs{UseBlob: true, Content: "http://127.0.0.1:1/blob"}
+	// Bind to 127.0.0.1:0 to claim a free port, then close the listener so
+	// any subsequent connection to that address is deterministically refused.
+	// This avoids depending on a specific port (e.g. tcpmux/1) being closed
+	// in CI/container environments.
+	ln, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatalf("listen: %v", err)
+	}
+	closedAddr := ln.Addr().String()
+	_ = ln.Close()
+	args := &common.CommandArgs{UseBlob: true, Content: "http://" + closedAddr + "/blob"}
 
-	_, err := h.fileUpload(args, er, 5, "blob.bin", false)
+	_, err = h.fileUpload(args, er, 5, "blob.bin", false)
 	if err == nil {
 		t.Fatal("fileUpload returned nil err, want PUT transport error")
 	}

